torch/utils/weak.py

```diff
@@ -0,0 +1,79 @@
+import weakref
+from collections.abc import MutableMapping
+from typing import Dict
+
+
+__all__ = ['WeakTensorRefKey', 'WeakTensorKeyDictionary']
+
+
+# Utility classes for working with weak references to tensors
+
+# torch.Tensors cannot be used as a key in a dictionary
+# because they define a custom __eq__ function which when used
+# to resolve hash collisions will throw when comparing tensors:
+# "RuntimeError: bool value of Tensor with more than one value is ambiguous."
+# To avoid that, we use an object which will hold a Tensor and use
+# its id for both hashing and equality.
+# In order to use this as a weak key reference, we cannot
+# simply use weakref.WeakKeyDictionary because the newly constructed
+# WeakTensorRefKey only use would be a dictionary so it would have no strong
+# references.
+# To get around this issue, we can use it as a normal key, and then set
+# `weakref.finalize` to delete the key when its contained tensor dies.
+
+
+class WeakTensorRefKey(object):
+    def __init__(self, ten):
+        self.ten = weakref.ref(ten)
+        # store id since as soon as ten is deallocated
+        # the old id will no longer be recoverable, and
+        # we need to be able to remove the WeakTensorRefKey
+        # from the dictionary by hashing it to the same
+        # value it had when ten was alive
+        self.id = id(self.ten())
+
+    def __hash__(self):
+        return self.id
```

Comment:
If you ever ever have a map or set or some instance where you need to hash the weakref alongside the real thing, this will be hard to figure out and debug. That and the comment above about memory validity makes me feel like we should just have another, non id unique here that we use to map 1:1

Reply:
IDK, if you want to setup a separate ID, you could possibly do so with a twinned object whose lifetime was the same as the original tensor, but has correct eq, and then use that as the hash, but... that seems more complicated for no gain?

Reply:
WDYM hash the weakref ? You can't use a hash the weakref of a Tensor. You would run into an error (thus this class).